data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -324,14 +329,24 @@ public void nonNullWrite(int rowId, LocalDateTime data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, BigDecimal data, ColumnVector output) {
-      // TODO: validate precision and scale from schema
+      Preconditions.checkArgument(data.scale() == scale,
+          "Cannot write value as decimal(%s,%s), wrong scale: %s", precision, scale, data);
+      Preconditions.checkArgument(data.precision() <= precision && precision <= 18,
```

Comment:
Nit: `precision <= 18` check can be moved into the constructor

Reply:
the `precision <=18 ` can be removed now, because we've checked it [here](https://github.com/apache/iceberg/pull/1271/files#diff-b1b07b15f036000a3f2bed76fdd9f961R108).